OpenRA.Mods.Common/FileSystem/ISO9660.cs

```diff
@@ -0,0 +1,226 @@
+#region Copyright & License Information
+/*
+ * Copyright (c) The OpenRA Developers and Contributors
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.IO;
+using System.Text;
+using OpenRA.FileSystem;
+using OpenRA.Primitives;
+using FS = OpenRA.FileSystem.FileSystem;
+
+namespace OpenRA.Mods.Common.FileSystem
+{
+	public class ISO9660Loader : IPackageLoader
+	{
+		public sealed class ISO9660Package : IReadOnlyPackage
+		{
+			public readonly record struct Entry(uint Offset, uint Length);
+
+			public string Name { get; }
+			public string VolumeName { get; }
+
+			public IEnumerable<string> Contents => index.Keys;
+
+			readonly Dictionary<string, Entry> index = [];
+			readonly Stream s;
+
+			readonly record struct DirectoryRecord(string Name, uint Offset, uint Length, bool IsDirectory);
+			static bool TryReadDirectoryRecord(Stream s, out DirectoryRecord record, bool isJoliet = false)
+			{
+				var start = s.Position;
+				var recordLength = s.ReadUInt8();
+				if (recordLength == 0)
+				{
+					s.Position = start;
+					record = default;
+					return false;
+				}
+
+				s.Position++;
+				var location = s.ReadUInt32();
+				s.Position += 4;
+				var length = s.ReadUInt32();
+				s.Position += 11;
+				var flags = s.ReadUInt8();
+				s.Position += 6;
+
+				var identifierLength = s.ReadUInt8();
+				var buffer = identifierLength < 128 ? stackalloc byte[identifierLength] : new byte[identifierLength];
+				s.ReadBytes(buffer);
+				var identifier = (isJoliet ? Encoding.BigEndianUnicode : Encoding.ASCII).GetString(buffer);
+
+				s.Position = start + recordLength;
+
+				record = new DirectoryRecord(identifier.Split(';')[0], 2048 * location, length, (flags & 2) != 0);
+				return true;
+			}
+
+			void EnumerateDirectories(DirectoryRecord parent, string path = null, bool isJoliet = false)
+			{
+				var pos = s.Position;
+				s.Position = parent.Offset;
+
+				// Skip . and .. records
+				TryReadDirectoryRecord(s, out _);
+				TryReadDirectoryRecord(s, out _);
+				while (s.Position < parent.Offset + parent.Length)
+				{
+					if (!TryReadDirectoryRecord(s, out var child, isJoliet))
+						break;
+
+					var childPath = path;
+					if (childPath == null)
+						childPath = child.Name;
+					else
+						childPath += "/" + child.Name;
```

Comment:
```suggestion
					var childPath = path == null ? child.Name : $"{path}/{child.Name}";
```

Maybe?

Also does ISO9660 always use "/" ?

Reply:
The standard doesn't specify a separator character, so we're choosing `/` for simplicity.